File: trabajo 14/bucket_sort.js
// Función de ordenamiento por inserción para una cubeta
function ordenamientoInsercion(cubeta) {
    for (let j = 1; j < cubeta.length; j++) {
        const valor = cubeta[j];
        let k = j - 1;

        while (k >= 0 && cubeta[k] > valor) {
            cubeta[k + 1] = cubeta[k];
            k--;
        }
        cubeta[k + 1] = valor;
    }
}

// Función principal del algoritmo Bucket Sort
function ordenamientoBucket(arreglo) {
    const tamaño = arreglo.length;
    const cubetas = Array.from({ length: tamaño }, () => []);

    // Distribuir elementos en cubetas
    for (const num of arreglo) {
        let indice = Math.floor(tamaño * num);
        if (indice === tamaño) indice = tamaño - 1; // Evita índice fuera de rango
        cubetas[indice].push(num);
    }

    // Ordenar cada cubeta
    for (const cubeta of cubetas) {
        ordenamientoInsercion(cubeta);
    }

    // Concatenar cubetas en el arreglo original
    let indiceArreglo = 0;
    for (const cubeta of cubetas) {
        for (const num of cubeta) {
            arreglo[indiceArreglo++] = num;
        }
    }
}


function main() {
    const arreglo = [0.88, 0.16, 0.48, 0.25, 0.75, 0.99, 0.23, 0.11, 0.24, 0.67];

    console.log("Arreglo desordenado:");
    console.log(arreglo.join(" "));

    ordenamientoBucket(arreglo);

    console.log("Arreglo ordenado:");
    console.log(arreglo.join(" "));
}


main();

File: trabajo 14/bucket_sort.cs
using System;
using System.Collections.Generic;

class Program
{
    // Método de ordenamiento por inserción para una cubeta
    static void OrdenamientoInsercion(List<float> cubeta)
    {
        for (int j = 1; j < cubeta.Count; j++)
        {
            float valor = cubeta[j];
            int k = j - 1;

            while (k >= 0 && cubeta[k] > valor)
            {
                cubeta[k + 1] = cubeta[k];
                k--;
            }
            cubeta[k + 1] = valor;
        }
    }

    // Función principal de Bucket Sort
    static void OrdenamientoBucket(List<float> arreglo)
    {
        int tamaño = arreglo.Count;
        List<List<float>> cubetas = new List<List<float>>(tamaño);

        // Inicializar cubetas vacías
        for (int i = 0; i < tamaño; i++)
        {
            cubetas.Add(new List<float>());
        }

        // Distribuir los elementos en cubetas
        foreach (float num in arreglo)
        {
            int indice = (int)(tamaño * num);
            if (indice == tamaño) // Evita índice fuera de rango
                indice = tamaño - 1;

            cubetas[indice].Add(num);
        }

        // Ordenar cada cubeta
        foreach (var cubeta in cubetas)
        {
            OrdenamientoInsercion(cubeta);
        }

        // Concatenar cubetas en el arreglo original
        int indiceArreglo = 0;
        for (int i = 0; i < tamaño; i++)
        {
            foreach (float num in cubetas[i])
            {
                arreglo[indiceArreglo++] = num;
            }
        }
    }

    static void Main()
    {
        List<float> arreglo = new List<float> { 0.88f, 0.16f, 0.48f, 0.25f, 0.75f, 0.99f, 0.23f, 0.11f, 0.24f, 0.67f };

        Console.WriteLine("Arreglo desordenado:");
        foreach (float num in arreglo)
        {
            Console.Write(num + " ");
        }
        Console.WriteLine();

        OrdenamientoBucket(arreglo);

        Console.WriteLine("Arreglo ordenado:");
        foreach (float num in arreglo)
        {
            Console.Write(num + " ");
        }
        Console.WriteLine();
    }
}
